util/mmio.h

```diff
@@ -206,11 +278,47 @@ __le64 mmio_read64_le(const void *addr);
    be in ascending address order.
 */
 #ifdef __s390x__
-static inline void mmio_memcpy_x64(void *dest, const void *src, size_t bytecnt)
+#define S390_PCI_MAX_WRITE_SIZE 128
+static inline uint8_t s390_get_max_write_size(uint64_t src, uint64_t dst, int len, int max)
 {
-	s390_mmio_write(dest, src, bytecnt);
+	int count = len > max ? max : len, size = 1;
+
+	while (!(src & 0x1) && !(dst & 0x1) && ((size << 1) <= count)) {
+		dst = dst >> 1;
+		src = src >> 1;
+		size = size << 1;
+	}
+	return size;
 }
 
+static inline void mmio_memcpy_x64(void *dst, const void *src, size_t bytecnt)
+{
+	int size;
+
+	if (unlikely(!s390_mio_supported))
+		s390_mmio_write(dst, src, bytecnt);
+
+	while (bytecnt > 0) {
+		size = s390_get_max_write_size((uint64_t __force) dst,
+					       (uint64_t) src, bytecnt,
+					       S390_PCI_MAX_WRITE_SIZE);
+		if (size > 8) /* main path */
```

Comment:
x64 is always chunks of exactly 64 bytes, all this logic seems impossible

Reply:
Ah, thanks, didn't realize this restriction. So you're saying the the bytecnt is always a multiple of 8, in that case yes we can greatly simplify this and get rid of the whole `s390_get_max_write_size()` thing.  I somehow read it as a general `mmio_memcpy()` and thought the x64 referred to the fact that it's only available on architectures with a 64 bit atomic.

Reply:
Not 8 bytes, 64 bytes. This is designed to be optimal for entire cache lines

Reply:
@jgunthorpe if this is designed for entire cache lines can I also assume that `dst` and `src` are 64 byte aligned?

Reply:
I can't remember,  think it is only 8 bytes required.

Reply:
Thanks

Reply:
Sorry for the confusion yesterday, I realized later that we still need another restriction currently ensured by `s390_get_max_write_size()` the `pcistbi` instruction can't normally write accross a 4096 byte boundary in the MMIO destination. So we will need some logic there.